### PR TITLE
feat: add tokenizers and estimate token use

### PR DESCRIPTION
**Pull request by Cifko**
Add tokenizers to the proxy for tokens estimation. Update the stacks with this.
During stack creation we have the event so we can update the DB right away.

### CODE REVIEW

File: atoma-proxy/src/server/mod.rs [hunk 1]
```diff
@@ -26,6 +27,8 @@ use crate::sui::Sui;
 const CHAT_COMPLETIONS_PATH: &str = "/v1/chat/completions";
 const MODELS_PATH: &str = "/v1/models";
 const NODE_PUBLIC_ADDRESS_REGISTRATION: &str = "/node/registration";
+const STACK_ENTRY_COMPUTE_UNITS: u64 = 1000;
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/mod.rs:30]: Can you add comments for all these constants ? The ones above, this one, and the ones below

File: atoma-proxy/src/server/mod.rs [hunk 2]
```diff
@@ -250,18 +336,40 @@ async fn get_selected_node(
             error!("No tasks found for model {}", model);
             return Err(StatusCode::NOT_FOUND);
         }
-        Ok(sui
+        // TODO: What should be the default values for the stack entry/price?
+        if total_tokens > STACK_ENTRY_COMPUTE_UNITS {
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/mod.rs:340]: Why do we have this check ? Total tokens can be as large as possible, if a node can process it.

File: atoma-proxy/src/server/mod.rs [hunk 3]
```diff
@@ -250,18 +336,40 @@ async fn get_selected_node(
...
+            return Err(StatusCode::BAD_REQUEST);
+        }
+        let event = sui
             .write()
             .await
-            .acquire_new_stack_entry(tasks[0].task_small_id as u64, 1000, 100)
+            .acquire_new_stack_entry(
+                tasks[0].task_small_id as u64,
+                STACK_ENTRY_COMPUTE_UNITS,
+                STACK_ENTRY_PRICE,
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/mod.rs:350]: why is this hardcoded ?

File: atoma-proxy/src/server/mod.rs [hunk 4]
```diff
@@ -250,18 +336,40 @@ async fn get_selected_node(
...
+            error!("Total tokens exceeds maximum limit of {STACK_ENTRY_COMPUTE_UNITS}");
+            return Err(StatusCode::BAD_REQUEST);
+        }
+        let event = sui
             .write()
             .await
-            .acquire_new_stack_entry(tasks[0].task_small_id as u64, 1000, 100)
+            .acquire_new_stack_entry(
+                tasks[0].task_small_id as u64,
+                STACK_ENTRY_COMPUTE_UNITS,
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/mod.rs:349]: why is this hardcoded ?

File: atoma-proxy/src/server/mod.rs [hunk 5]
```diff
@@ -278,36 +386,40 @@ pub struct NodePublicAddressAssignment {
...
+              "id": model,
+              "object": "model",
+              "created": 1730930595,
+              "owned_by": "vllm",
+              "root": model,
+              "parent": null,
+              "max_model_len": 2048,
+              "permission": [
+                {
+                  "id": format!("modelperm-{}", model),
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/mod.rs:406]: Where is the metadata below coming from ?

File: atoma-proxy/src/server/mod.rs [hunk 6]
```diff
@@ -278,36 +386,40 @@ pub struct NodePublicAddressAssignment {
...
-        ]
+        "data": state
+        .models
+        .iter()
+        .map(|model| {
+            json!({
+              "id": model,
+              "object": "model",
+              "created": 1730930595,
+              "owned_by": "vllm",
```
Comment (jorgeantonio21) [inline atoma-proxy/src/server/mod.rs:400]: why `vllm` ?

File: atoma-state/src/handlers.rs [hunk 7]
```diff
@@ -31,11 +32,9 @@ pub async fn handle_atoma_event(
             handle_node_task_unsubscription_event(state_manager, event).await
         }
         AtomaEvent::StackCreatedEvent(event) => {
-            if event.owner == state_manager.state.owner {
-                handle_stack_created_event(state_manager, event).await
-            } else {
-                Ok(())
-            }
+            // Don't handle creation here. It's handled when the stack is created right away.
```
Comment (jorgeantonio21) [inline atoma-state/src/handlers.rs:35]: Makes sense, please change it to:

`NOTE: Don't handle creation here. It's handled when the stack is created right away.`

File: atoma-state/src/state_manager.rs [hunk 8]
```diff
@@ -260,12 +255,23 @@ impl AtomaState {
     pub async fn get_stacks_for_model(&self, model: &str, free_units: i64) -> Result<Vec<Stack>> {
```
Comment (jorgeantonio21) [inline atoma-state/src/state_manager.rs:255]: Can we add unit tests for the update part ?

**Review by jorgeantonio21 — COMMENTED** [review]
LGTM, 

left a few minor comments to be addressed.

File: atoma-proxy/src/server/mod.rs (hunk 1 above)
Comment (Cifko) [reply atoma-proxy/src/server/mod.rs:30]: Added todos

File: atoma-proxy/src/server/mod.rs (hunk 3 above)
Comment (Cifko) [reply atoma-proxy/src/server/mod.rs:350]: it should not be, explained above.

File: atoma-proxy/src/server/mod.rs (hunk 4 above)
Comment (Cifko) [reply atoma-proxy/src/server/mod.rs:349]: same as above

File: atoma-proxy/src/server/mod.rs (hunk 5 above)
Comment (Cifko) [reply atoma-proxy/src/server/mod.rs:406]: This is what vllm returned and I just used it. I have to check what this structure should/can.
Added todo.

File: atoma-proxy/src/server/mod.rs (hunk 6 above)
Comment (Cifko) [reply atoma-proxy/src/server/mod.rs:400]: changed to `atoma` :-)

File: atoma-state/src/handlers.rs (hunk 7 above)
Comment (Cifko) [reply atoma-state/src/handlers.rs:35]: done

File: atoma-state/src/state_manager.rs (hunk 8 above)
Comment (Cifko) [reply atoma-state/src/state_manager.rs:255]: The problem is that for testing you will need to run postgres. That's why there are no tests in this file.

**Pull request merged**